Docs: Bug Fixes for API Documentation Pages

jperson2000 opened the pull request:
## Description

This update fixes bugs mentioned in #9273 and #9312 for the API documentation pages and adds 45 tests in 3 bUnit test classes to exercise docs components:

#### Reported by @iwis:
- [X] Re-add the "Categories/Inheritance/None" grouping for members
- [X] The "Events" section (for events and `EventCallback`) should be present
- [X] `see cref` links should not be blank when linking to an external type such as `EditForm`
- [X] Some incorrect descriptions: `MudChipSet<T>.Comparer`, `MudHighlighter.Markup`, `MudMessageBox.YesButton`
- [X] There are no descriptions for many properties, such as `MudColorPicker.Adornment`.
- [X] When grouping by category, sort by the category Order, then by member Name.  
- [X] In mobile mode, API text should wrap properly

#### Reported by @danielchalmers:
- [X] Use a `max-width` on columns (when not in mobile mode) to prevent oversized columns.
- [X] UI is blocked while loading APIs (should be able to type or scroll)
- [X] Documented* classes may want to be immutable with no expression properties  **(JP: Changed to `init`, removed some expression properties.)**
- [X] External sources (MSFT docs) could have an icon indicating they open in a new tab. **(JP: Added MudTooltip and "link" icon)**
- [X] Padding could be reduced in docs-section-content-inner to show more data on screen at once
- [X] The "visit the example page" link says "MudAppBar" but the page title is "App Bar"
- [X] Change title summary Typo from `caption` to `subtitle`
- [X] Show icons for external links so users know what's external.  Show the icons on the right of the link.
- [X] Truncate the tooltips so they aren't too long.
- [X] The subtitle text under the component name wraps if there's a link in the text.  
- [X] In mobile mode, API text should wrap properly

#### Reported by @ScarletKuro:
- [X] Internal properties like `ParameterContainer`, `InputIdState` should not be included
- [X] Show an icon for two-way bindable parameters (i.e. property paired with an `EventCallback` when changed)
- [X] Change to `LoxSmoke.DocsXml` for back-end generation
- [X] Remove the compiler warnings for undocumented members

#### Reported by @igotinfected: 
"API" section navigation disappears (see below)

#### Reported by @mikes-gh:
Use a source generator (see below)

#### Reported by @jperson2000:
- [X] ApiLink throws exception if an XML element is self-closing (e.g. `<see />`)
- [X] Show method return types and parameters for the "Methods" table.
- [X] When grouping by "None", set the GroupBy to `null`

### New & Restored Features

#### Inheritance Tree
![image](https://github.com/user-attachments/assets/913279ed-3efd-4d3a-b089-d19f6e910a84)

#### (Restored) Grouping and Keyword Search
![image](https://github.com/user-attachments/assets/7e7faefd-0f98-47f5-99db-c0a46c3a017b)

#### External Link Tooltips with Icon
![image](https://github.com/user-attachments/assets/55267528-2d2f-4ca8-97a4-4b8d8db9a011)

#### Icon and Tooltip for Bindable Properties and Events
![image](https://github.com/user-attachments/assets/e09c40ed-e759-4a03-95ac-5853c6362aca)

![image](https://github.com/user-attachments/assets/0d97be98-cf44-494e-9d47-e4d94fba9163)

#### Tooltips for Each Method Parameter
![image](https://github.com/user-attachments/assets/a3cc07a1-dd9e-403f-990b-81472df0ce3c)

#### Performance Profile of MudBlazor.Docs.Compiler

![image](https://github.com/user-attachments/assets/1fdefd21-01db-4336-85c2-3ce9ab580039)

#### Reported Issues I Was Unable to Fix
:exclamation: Linking to a member shouldn't show the member off-screen.  **(JP: This requires `MudTable.RowTemplate` to have some `RowId` function to provide an anchor)**
:exclamation: Overall memory and cpu usage is increased  **(JP: Source-generated API pages would solve this after bugs are fixed, by moving the data to each individual page instead of one massive dictionary.)**
:exclamation: Evaluate MenuService,ApiLinkService,ApiDocumentation for refactor/merge  **(JP: Used MenuService to get component name)**
:exclamation: Use a Source Generator instead of `ApiDocumentationWriter` to generate doc objects during builds.  **(JP: In the next PR, perhaps we can source-generate Razor pages for each type (e.g. "[TypeName].razor") and greatly reduce memory)**
:exclamation: The previous/next links at the bottom of an API page say "App Bar" but the page title is "MudAppBar"  **(JP: I tried to fix this but was unsuccessful)**
:exclamation: When clicking a link to go to a component page from the API page, the "API" link disappears.  **(JP: This one is tricky due to the design of `SectionHeader` and it preserving state across pages)**
:exclamation: If there's only one category ("General") for a section ("Methods") it could skip the header altogether  **(JP: Not sure how to do this, sorry!)**
:exclamation: When changing pages, the section navigation on the right often disappears.  **(JP: I tried to fix this but was unsuccessful)**

## How Has This Been Tested?

This was tested by observing the API documentation pages for all components and observing all tests.

## Type of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation (fix or improvement to the website or code docs)

## Checklist

- [X] The PR is submitted to the correct branch (`dev`).
- [X] My code follows the code style of this project.
- [X] I've added relevant tests.

jperson2000 committed:
API Docs: Restored Grouping buttons: None/Categories/Inheritance

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

jperson2000 committed:
Switched to LoxSmoke.DocXml for back-end generation of XML docs.  Fixed ApiLink bug regarding self-closed elements.  Excluded some internal types and members from API docs.

jperson2000 committed:
Merge branch 'feature/xmldocs-bugfixes' of https://github.com/jperson2000/MudBlazor into feature/xmldocs-bugfixes

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

jperson2000 committed:
Improved support for events and bindable properties.

jperson2000 committed:
Merge branch 'feature/xmldocs-bugfixes' of https://github.com/jperson2000/MudBlazor into feature/xmldocs-bugfixes

jperson2000 committed:
Added warmup during app startup to load API docs

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

jperson2000 committed:
Added support for methods, return value docs, and method parameter docs.

jperson2000 committed:
Merge branch 'feature/xmldocs-bugfixes' of https://github.com/jperson2000/MudBlazor into feature/xmldocs-bugfixes

jperson2000 committed:
Fixed issues when linking to external types and members.  Fixed slight XML docs error in MudBaseButton.

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

jperson2000 committed:
dotnet format

jperson2000 committed:
Merge branch 'feature/xmldocs-bugfixes' of https://github.com/jperson2000/MudBlazor into feature/xmldocs-bugfixes

jperson2000 committed:
Fixed padding of docs sections for API pages.  Added tooltips for method parameters.

danielchalmers commented:
Looking good so far! Could you change the text below the header from `caption` to `subtitle1` like the component page?

![image](https://github.com/user-attachments/assets/2a3af034-28af-4058-8709-d39cb4eec41e)

And add a break in the alert so the link itself is easier to scan visually

![image](https://github.com/user-attachments/assets/8847620f-f17f-4ce6-8bd5-e58b79b54d35)

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

jperson2000 committed:
Added tooltips and link icons for external links.  Added Typo parameter for ApiText.  Change main API doc text from Typo to Subtitle.  Ready to review!

jperson2000 committed:
Merge branch 'feature/xmldocs-bugfixes' of https://github.com/jperson2000/MudBlazor into feature/xmldocs-bugfixes

jperson2000 commented:
Okay @henon @danielchalmers @ScarletKuro @igotinfected @mikes-gh this is now ready to review!  I've fixed the majority of bugs, changed the back end to use `LocSmoke.XmlDocs` (which greatly increased the amount of documentation shown), and I think it's looking good.   I have free time for the next several days to address any issues, and thanks as always for taking the time to review!

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

jperson2000 committed:
Fixed accidental infinite loop.  All tests pass!

jperson2000 committed:
Merge branch 'feature/xmldocs-bugfixes' of https://github.com/jperson2000/MudBlazor into feature/xmldocs-bugfixes

jperson2000 commented:
Okay @henon @danielchalmers @ScarletKuro @igotinfected @mikes-gh this is now ready to review!  I've fixed the majority of bugs, changed the back end to use `LocSmoke.XmlDocs` (which greatly increased the amount of documentation shown), and I think it's looking good.   I have free time for the next several days to address any issues, and thanks as always for taking the time to review!

danielchalmers commented:
Personally appreciate the dedication to the docs! Only noticed small style things:

If you go from an API page to the component page it won't show this part:

![image](https://github.com/user-attachments/assets/4c362855-2d21-4d93-8e28-6df77de299ba)

Minor request but could the icon be on the right for alignment?

![image](https://github.com/user-attachments/assets/8318e44d-eefa-4ea5-b40e-6e14a8c3fe65)

I'd also say it doesn't need the tooltip because you can see it in the browser itself and the icon itself could have a sole tooltip "External Link":

![image](https://github.com/user-attachments/assets/89338f97-d13f-458f-9710-3a3486a47579)

![image](https://github.com/user-attachments/assets/6e5c9c50-65a8-4161-a5cf-bbe45cfca889)

Maybe all the tooltips should have a max length before truncation so they don't fill up too much of the screen:

![image](https://github.com/user-attachments/assets/4894a39c-f1cd-4410-b45a-fc1dd02cfac9)

e.g. `The color themes available in MudBlazor, allowi...`

I see a stray period due to eager wrapping:

![image](https://github.com/user-attachments/assets/48c0dd87-a27f-41c8-bdae-ba501ab4f88a)

iwis commented:
You changed the order of groups of properties to alphabetical once again. The order is semantic on the production and development servers. https://github.com/MudBlazor/MudBlazor/discussions/9077

jperson2000 committed:
Fixed category sort issue reported by @iwis

jperson2000 committed:
Fixed tooltips for external links reported by @danielchalmers.  Fixed sorting when using "None" grouping.

jperson2000 commented:
> You changed the order of groups of properties to alphabetical 

I appreciate it, @iwis!  I think I finally have this correct and thanks for taking the time to review.

jperson2000 commented:
> Minor request but could the icon be on the right for alignment?

Thanks as always for the great suggestions @danielchalmers !  The icon is now on the right side for external links.

> I'd also say it doesn't need the tooltip because you can see it in the browser itself and the icon itself could have a sole tooltip "External Link":

The link icon (and just the icon) now shows "External Link" (without the URL) as a tooltip.

> Maybe all the tooltips should have a max length before truncation so they don't fill up too much of the screen:

I capped the maximum tooltip length at 60, after which it truncates it and adds an ellipsis...

> I see a stray period due to eager wrapping:

This started happening when the `Typo` was changed to `Subtitle`, so I'm not sure how to fix this without modifying the style for other people.  Do you have any ideas for a workaround?

iwis commented:
> > You changed the order of groups of properties to alphabetical
> 
> I appreciate it, @iwis! I think I finally have this correct

I still don't see this.

On https://mudblazor.com/api/MudCheckBox%601 the order is semantic:
  General, Data, Validation, Behavior, Appearance, Common.

On https://localhost:5001/api/MudCheckBox%601 the order is alphabetical:
  Appearance, Behavior, Common, Data, General, Validation.

It should be like on the production server (but the `General` category should be at the end because it means the category wasn't set for the property):
  Data, Validation, Behavior, Appearance, Common, General.

> thanks for taking the time to review

I haven't done a review yet because I am unable to do it until there are major inconsistencies between https://v6.mudblazor.com and the current version. When there are no major inconsistencies, I would like to programmatically or manually compare the output of both versions on all API pages and find minor differences potentially resulting from less visible bugs. A few months ago, it would have been easy because there were no differences in the content. It can be harder now and even impossible in a few months if the MudBlazor API will change more.

jperson2000 committed:
Fixed wrapping issue.  Thanks @danielchalmers!

jperson2000 commented:
> I think you can change the HtmlTag on MudText to `span` to fix that because `subtitle1` uses a paragraph tag which takes up an entire line

Thank worked perfectly!  Daniel to the rescue!

![image](https://github.com/user-attachments/assets/045a69e1-efde-42a8-87fc-bd68d655738a)

jperson2000 commented:
> It should be ... (`General` category should be at the end because it means the category wasn't set for the property): Data, Validation, Behavior, Appearance, Common, General.

That's what I'm currently seeing, yes, where **General** is at the end.  Technically speaking: If a category hasn't been specified via `CategoryAttribute`, a member is assigned `int.MaxValue` to the `Order` making them sort to the bottom of the list.

iwis commented:
> That's what I'm currently seeing, yes.

Oh! Thanks. Now I see it also. Probably I haven't refreshed something.

jperson2000 commented:
> Oh! Thanks. Now I see it also. Probably I haven't refreshed something.

All good, I love your attention to detail and appreciate your hard work on these pages!  Great stuff.

iwis commented:
And a similar topic - what do you think about names of groups of properties and their order when you click on the "Inheritance" button in the "None|Category|Inheritance" switch?

On https://v6.mudblazor.com/api/checkbox the order of groups is semantic according to the inheritance hierarchy:
 1. the group without the header - means that these properties are not inherited
 2. "Inherited from MudBooleanInput<T>"
 3. "Inherited from MudFormComponent<T, bool?>"
 4. "Inherited from MudComponentBase"

On https://localhost:5001/api/MudCheckBox%601 the order of groups it is alphabetical:
 1. "MudBooleanInput<T>"
 2. "MudCheckBox<T>"
 3. "MudComponentBase"
 4. "MudFormComponent<T, U>"

Is the alphabetical order better?
IMHO, also adding the "Inherited from " prefix in the names of headers is more informative.

jperson2000 committed:
Improved Inheritance grouping thanks to @iwis

jperson2000 commented:
> And a similar topic - what do you think about names of groups of properties and their order when you click on the "Inheritance" button in the "None|Category|Inheritance" switch?

That is a much better way to sort :eyes: great idea!   Ok, I have made that improvement.  

> IMHO, also adding the "Inherited from " prefix in the names of headers is more informative.

Which do you think is best: an empty group, or one with the current class, or the current class with " (This Class)" appended?

**Option 1: Empty Group**
![image](https://github.com/user-attachments/assets/53e6aa92-36c6-4811-9860-bf990f1057c6)

**Option 2: Just the Class Name**
![image](https://github.com/user-attachments/assets/19a7e8e0-71ea-456e-b813-a72469d43da7)

**Option 3: Class Name with " (This Class)"**
![image](https://github.com/user-attachments/assets/43d11a36-d12c-425b-8045-131fb557cd6b)

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

iwis commented:
I believe that Option 1 (empty) can be the most elegant and simple. Probably we should also consider elements other than MudBlazor component classes - e.g. enumerations like https://localhost:5001/api/Position - they have no inheritance so no header is better because it is simpler.  But display only one horizontal line not two.
Option number two is also fine.
Option number 3 isn't good because enums are not classes, so "Position (This Class)" on https://localhost:5001/api/Position will be incorrect.

jperson2000 committed:
Changed to "Option 1" (No group name for current type) when grouping by inheritance.  Thanks @iwis

jperson2000 commented:
> I believe that Option 1 (empty) can be the most elegant and simple. Probably we should also consider elements other than MudBlazor component classes - e.g. enumerations 

Excellent point!  Ok, Option 1 (Empty) has been committed.

jperson2000 committed:
Merge branch 'MudBlazor:dev' into feature/xmldocs-bugfixes

jperson2000 committed:
Removed some unneeded code

iwis commented:
I see that you show protected properties that are not parameters, for example, `MudTextField.Classname`. They were not shown in the v6. Is it deliberate? If yes, then what is the reason for this decision?

Compare:
  https://localhost:5001/api/MudTextField%601 (the `Classname` property is present)
  https://v6.mudblazor.com/api/textfield            (the `Classname` property  is absent)

iwis commented:
If it was not discussed previously, just don't show them. As far as I remember, in v6 there was a code that only parameters were displayed, i.e. only properties with the `Parameter` attribute - but please check it in the source code.

iwis commented:
Yep, as I said it was: `Type.GetPropertyInfosWithAttribute<ParameterAttribute>()` in https://github.com/MudBlazor/MudBlazor/blob/v6/src/MudBlazor.Docs/Components/DocsApi.razor#L211

iwis commented:
There is also `MudTextField<T>.InputReference`, which is a public property but doesn't have the parameter attribute. Is it useful?

iwis commented:
Similarly, the method `internal override InputType GetInputType()` is displayed now in MudTextField, but not in v6.

iwis commented:
I checked properties in the "General" category on https://localhost:5001/api/MudTextField%601. Most of them are public properties not being parameters. They shouldn't be displayed or they should have a proper category added.

ScarletKuro commented:
> Most of them are public properties not being parameters. They shouldn't be displayed or they should have a proper category added

hey should be shown but categorized, because these can be accessed with `@ref`, and we often find that this is what the customer wants.

iwis commented:
> They should be shown but categorized,
Thank you for the info. I don't use Blazor, so don't know some crucial things (I am still waiting for it to gain popularity / conquer the world ;) ). So they should be displayed and categorized. Then another question to consider: Maybe they should differ in some way from parameters? Maybe some icon, or another category, or another section apart from Parameters, MEthods and Events, or something else. IDK. I will be back in a few hours, need to go now.

danielchalmers reviewed (approved):
Really solid work!

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

jperson2000 commented:
> > Most of them are public properties not being parameters. They shouldn't be displayed or they should have a proper category added
> 
> They should be shown but categorized, because these can be accessed with `@ref`, and we often find that this is what the customer wants.

Thanks a lot for the discussion, guys.  What if the member tables had a `ShowProtected` parameter, and a `MudSwitch` to control it, which defaults to `false`?   This would hide protected members by default but give users an option to show them.

ScarletKuro commented:
> What if the member tables had a `ShowProtected` parameter

That sounds good to me.

What I would do, to be honest (not now, but in the future), is create three sections:

1. Component parameters ordered by category priority. These are parameters that can be used declaratively in the Razor syntax.
2. Public properties/methods for imperative use (`@ref`), ordered alphabetically.
3. Protected properties/methods for inheritance (composition use), ordered alphabetically.

This makes the most sense to me.

jperson2000 committed:
Added support for filterable protected members per @iwis and @ScarletKuro

jperson2000 commented:
> > What if the member tables had a `ShowProtected` parameter
> 
> That sounds good to me.

Ok, I've added a `MudSwitch` for `Show Parameters`, but only if there actually are protected members.  It is disabled and hides protected members by default. When enabled, protected members are shown with a "lock" icon, like this:

If there are protected members:
![image](https://github.com/user-attachments/assets/937a6c15-046c-48b5-bcc2-538d7cfa5090)

If there are no protected members:
![image](https://github.com/user-attachments/assets/f9e992ad-eb78-4106-84ad-70108b32584d)

Icon and Tooltip for protected members
![image](https://github.com/user-attachments/assets/912603c0-fe62-49e6-b708-7808fcfd766e)

We do have the `DocumentedMember.IsParameter` flag available, so it would not be hard to separate parameters from non-parameters in the next PR.

jperson2000 committed:
Resolved conflicts

jperson2000 committed:
Added tests for ApiText. (More to come.) Fixed some bugs found during testing.  Added optimizations for ApiDocumentation speed.

jperson2000 committed:
Added tests for ApiTypeLink.

jperson2000 committed:
Improved comments for tests

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

iwis commented:
The API doc isn't responsive. It should, because it is a component library. The v6 version was much more responsive.

iwis commented:
The API doc isn't responsive. It should be because it is a component library. The v6 version was much more responsive.

Look for example for the `Pattern` or `Validation` properties on the page https://localhost:5001/api/MudNumericField%601 on a smartphone (or press F12 in Google Chrome, show the device toolbar, and select the Samsung Galaxy S20 Ultra or Samsung Galaxy S8+). I think programmers don't check API too often on smartphones, but it isn't good if the documentation of a component library isn't responsive.

![image](https://github.com/user-attachments/assets/d2b44055-9e33-4cff-b7f4-4de5b349be85)

jperson2000 committed:
Added tests for ApiMemberTable.  Sealed non-component classes for speed.  Fixed ApiText wrapping on mobile.

jperson2000 committed:
Merge branch 'feature/xmldocs-bugfixes' of https://github.com/jperson2000/MudBlazor into feature/xmldocs-bugfixes

jperson2000 committed:
Improved toolbar for mobile mode.

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

jperson2000 committed:
Fixed two minor test issues.

ScarletKuro commented:
Where did the message about responsiveness go?

iwis commented:
> Where did the message about responsiveness go?

I moved it here: https://github.com/MudBlazor/MudBlazor/issues/9273#issuecomment-2426164074. I thought that as the title of the issue https://github.com/MudBlazor/MudBlazor/issues/9273 says it may be a good place to add all issues related to the API doc that were introduced by Jon's improvement. Also, I didn't report responsiveness previously, so didn't want to overload this PR.

jperson2000 commented:
> > Where did the message about responsiveness go?
> 
> I moved it here

I've been updating the PR and topic with summarized versions of people's reports as well.   

- [X] In mobile mode, API text should wrap properly

This should be looking better as of yesterday's commits.  Anything else I should get into this PR?

iwis commented:
@ScarletKuro wrote that internal properties should not be included, but you still display internal methods like MudTextField.GetInputType(). As far as I remember, internal members cannot be accessed by end users, so they should not be displayed.

Apart from this, I don't see other major issues that I know how to solve.

jperson2000 commented:
> @ScarletKuro wrote that internal properties should not be included

Okay, I will look into those tonight.

> Apart from this, I don't see other major issues that I know how to solve.

@iwis, I am really grateful that you took so much time to review and provide suggestions!

iwis commented:
> I am really grateful that you took so much time to review and provide suggestions!

Thank you for your good words, and thank you for your commitment to fix bugs.

jperson2000 committed:
Excluded all internal members from documentation generation.  Fixed null reference exception when grouping by Inheritance and a member is declared in an external (non-MudBlazor) type such as ComponentBase.

jperson2000 commented:
> @ScarletKuro wrote that internal properties should not be included

Ok, I think this is now fixed.  The generator will now exclude any member marked as `internal`  (`IsAssembly` = `true` in reflection).  I think we're in good shape now, but I'll stand by if anyone wants any other revisions.  Thanks again for everyone's detailed bug reports!

jperson2000 committed:
Merge branch 'dev' into feature/xmldocs-bugfixes

jperson2000 committed:
Fixed issue where some members had no DeclaringType set.  Thanks to @iwis once again.

jperson2000 commented:
Ok, I found the issue with `MudTimePicker` and fixed the issue.  I now see proper display and sorting of base types when sorting by **Inheritance**: 

![image](https://github.com/user-attachments/assets/e8aa8734-1b90-4fb7-9c4a-6192c6cb839b)

I think we're caught up on the issues which can be fixed.  Thanks again @iwis!

henon reviewed (approved):
Wow, even test for the Docs API, what a great PR! @ScarletKuro you can merge if you think it ready.

ScarletKuro commented:
@jperson2000 do you want address the responsive problems in separate pr or on this PR?
If separately then I'm more than happy to merge it now.

jperson2000 commented:
> @jperson2000 do you want to address the responsive problems in separate pr or on this PR? If separately then I'm more than happy to merge it now.

I think I'd prefer to merge this since it has important fixes, but then to continue with iwis's requests.  Thanks!
